### ISSUE TITLE
[DismissableLayer] Layering breaks with different component versions

### ISSUE THREAD

**Issue by jjenzz** [radix-ui/primitives#1088]
Initially raised [on Discord](https://discord.com/channels/752614004387610674/803656530259738674/931602628327145483).

If two components (e.g. `DropdownMenu` and `Dialog`) are using different versions of `DismissableLayer` internally, then the layering breaks because of the different context instances.

We never added scoping to `DismissableLayer` since it should always cross component boundaries, however, I wonder if scoping could help here since it effectively passes context through a private prop that it expects the component to use. It might be worth looking into as a solution.

**Comment by benoitgrelard** [comment]
Hmmm yeah that's where these multi-package distributions become complicated sometimes, these types of issues are really hard to stay on top of.

I'm worried the scoped context would kinda break the intention of `DismissableLayer` though.

**Comment by jjenzz** [comment]
could this be solved by changing the `"@radix-ui/react-dismissable-layer": "workspace:*"` dependency in each of the `package.json`s to `"@radix-ui/react-dismissable-layer": "latest"` ? 🤔

**Comment by benoitgrelard** [comment]
I don't think that would help as ultimately this is already the case, when a new change is made in `@radix-ui/react-dismissable-layer` the other packages would have a new version with the version of `@radix-ui/react-dismissable-layer` being latest. But it doesn't prevent anybody from installing versions that use different version of `@radix-ui/react-dismissable-layer`. Hardcoding to `"latest"` instead of a said version would kinda solve that, but I'm not sure it's viable or even makes sense from a code perspective as potentially changes could be introduced and older versions of popover for example might not be compatible with those changes forced by the latest.

Interestingly though, I had a somewhat similar thought process a few weeks back. I was thinking potentially we could "solve" this issue by making the incompatibility visible at install time by making `@radix-ui/react-dismissable-layer` a peer dependency.

**Comment by jjenzz** [comment]
> Hardcoding to `"latest"` instead of a said version would kinda solve that, but [...] changes could be introduced and older versions of popover for example might not be compatible with those changes forced by the latest.

i had considered this but since `DismissableLayer` is a "global" component I had thought maybe changes to its apis could be handled withy deprecation warnings and things. given that it's a pretty stable comp at this point and not likely to change much, that might not be so bad. it seems better than the current state of things anyway where things just completely break for folk.

making it a peer dependency would surely suffer the same issues except radix would be adding an extra step for consumer by requiring them to manage the dependency.

**Comment by pedromcunha** [comment]
Hi, just want to start out by saying thank you for building such a useful and headless primitive UI library. It really had helped our team when it comes to building with speed and reliability. I just wanted to throw in our scenario into the mix. We currently have an open source library that utilizes radix to build out the UI. I think it's ok for us to make sure that the dependencies are all on the same version (especially for those that use `DismissableLayer`). The issue arises when others install our library into their react application and they so happen to be using radix internally. I think at the moment it's not likely that this will happen too often but it would be nice in those cases to specify that we only allow using a specific version of `DismissableLayer`. In those cases either a warning or an error would suffice. 

I also tried adding `"@radix-ui/react-dismissable-layer": "1.0.1-rc.2"` as a peerDependency and this did warn me when installing into our demo application but after installing that I was left with two versions of `react-dismissable-layer`. We'll add this to our docs to warn users but it does make things a bit more complicated when someone wants to use our library and radix primitives. Maybe a better approach would be to detect it at run time and throw an error in the console when multiple `DismissableLayer` contexts are detected.

**Comment by benoitgrelard** [comment]
> > Hardcoding to `"latest"` instead of a said version would kinda solve that, but [...] changes could be introduced and older versions of popover for example might not be compatible with those changes forced by the latest.
> 
> i had considered this but thought maybe changes to its apis could be handled with a deprecation warning. given that it's a pretty stable comp at this point and not likely to change much, that might not be so bad. it seems better than the current state of things anyway where things break for folk.

I'm more worried about dependency changes, whilst the API of `DismissableLayer` might not change much, even a small change in any underlying dep (`@radix-ui/primitive`, `@radix-ui/react-primitive` for example) will generate a new version which will bring the issue again.

> something like:
> 
> > A component is using a deprecated `DismissableLayer` api. To solve this, upgrade your components to the same version number.

Also on this, "the same version number" is very hard to communicate as they're independently versioned so unlikely to be aligned, so effectively you have to say, make sure the primitives that use `DismissableLayer` are all aligned in terms of its version. It's very difficult… :(

**Comment by jjenzz** [comment]
> even a small change in any underlying dep (`@radix-ui/primitive`, `@radix-ui/react-primitive` for example) will generate a new version which will bring the issue again.

ofc, but what i'm saying is that radix would know that those sorts of changes will be **the only scenario** that will cause this issue for consumers so you can add a note to release notes / code warnings on with a migration path when you make changes to the underlying deps/apis (which again, its deps are pretty minimal/stable so rarely likely to change).

at the moment, the current state of things means that this could happen any time for any consumer just by changing their configuration of dependencies which radix has **zero** control over.

i agree that the suggestion isn't perfect but it seems better than the current alternative, and is a relatively simple change.

**Comment by maxichrome** [comment]
For all having this issue _now_:

Our current solution to this is using pnpm's [dependency override](https://pnpm.io/package_json#pnpmoverrides) functionality.

Adding this to `package.json` forces everything to use DismissableLayer version 1.0.2.

```json
"pnpm": {
	"overrides": {
		"@radix-ui/react-dismissable-layer": "1.0.2"
	}
},
```

**Comment by ysulyma** [comment]
@benoitgrelard 
> The only way to resolve this would be to use some global on `window` probably, which wouldn't be a good idea for many reasons

Don't know if this will work for you, but I solved this in my library by using a Symbol: https://github.com/liqvidjs/plugins/blob/4f19efbd15399cecc1721142936b4f191de52d70/packages/playback/src/react.ts#L6-L14

**Comment by StephenHaney** [comment]
I wonder if we can detect this and throw loudly in dev? The possibility that installing a new primitive silently creates a serious production bug is concerning (`pointer-events: none` stuck on body element).

My scenario was a Select inside of a Dialog... with both open, pressing escape closed both and `pointer-events: none` was stuck on the body element. I originally thought it was a bug in my code and spent a while debugging. It can be tricky to debug with dev servers and npm quirks, lots of variables.

Select and Dialog were both updated to the newest package version, but some of the other primitives were not. Once I updated everything the issue was resolved.

**Comment by StephenHaney** [comment]
[Here's how mobx handles cross-environment global](https://github.com/mobxjs/mobx/blob/main/packages/mobx/src/utils/global.ts) for the purpose of detecting multiple versions of the package at runtime. Basically use `globalThis` where it's supported and then fall back to individual environment global names. `react-dismissable-layer` could register itself to a global in this way and, if it's already there, throw an error.

The `peerDependencies` route seems promising as well but I think that would require the app author to install `react-dismissble-layer` manually in addition to the primitives?

**Comment by cowjuh** [comment]
This worked for me:

1. Overriding the `react-dismissable-layer` version in `package.json`

```
  "resolutions": {
    "@radix-ui/react-dismissable-layer": "^1.0.5"
  }
```

2. Deleting and reinstalling `node_modules`

**Comment by sean-luca** [comment]
The above solution also worked for me, but I had to include the dialog library as well:

```
"resolutions": {
    "@radix-ui/react-dismissable-layer": "^1.0.5",
    "@radix-ui/react-dialog": "^1.0.5"
  }
```

And +1 for deleting `node_modules` and re-installing.

**Comment by joshuaellis** [comment]
We're facing this issue in https://github.com/strapi/strapi. We use radix as a bedrock for https://github.com/strapi/design-system but because we're extensible there appears to be some conflicts happening in specific projects.

We're looking to recommending using resolutions as a workaround – but is there any appetite to investigate into solutions further? I appreciate it's very difficult & I'm not gonna pretend I know where to start in solving it either 😅

**Comment by StephenHaney** [comment]
Yes, definitely appetite to work on this! There isn't a straightforward answer but this one is very much top of mind for me. We're going to take another look. 

Looking for opinions on this: At minimum, I'd like this situation to break loudly in development, so you at least know you have mismatched versions.

**Comment by jjenzz** [comment]
@StephenHaney the `DismisssableLayer` props and context shape (external interface) hasn't changed in 2+ years. my original proposal here that was to hardcode the `"latest"` string into each package for their `react-dismissable-layer` dependency wld have solved this quickly and easily.

we're perhaps overthinking how imperfect this solution might be. while i agree it isn't ideal it wld have at least solved it for the past 2+ years while the team thinks on it some more.

it would allow your proposal of loudly breaking in development bcos when `DismissableLayer` context shape/props change, it can check if something is still using a legacy version of those interfaces and scream. they wouldn't know **which** component is relying on an old API but it can propose upgrading them all to the same version.

the benefit of this solution is that **it would only ever break loudly if the `DismissableLayer` interface changes**, which again, that hasn't happened in 2+ years 😛

**Comment by StephenHaney** [comment]
@jjenzz I completely agree, I think we need to take a pragmatic approach here and remove the trap.

How do you contrast your solution against using a `global this` type solution – which is actually pretty reliable across envs now – to detect multiple versions of `DismissableLayer` and throw in dev? So basically keeping a global that DismissableLayer registers itself onto (this is what mobx does).

**Comment by jjenzz** [comment]
@StephenHaney the difference is that a `globalThis` solution means they will have different versions of `DismissableLayer` in their bundle still (bundle size) and the chances of someone hitting this issue are the same as they are currently, we've just improved the messaging.

radix could remove the `DismissableLayerContext` altogether and `useSyncExternalStore` so all packages reference the same version of global state, however, that'd just **move** where breaking changes can occur—instead of them occurring when the consumer API changes (my proposal) they would occur when an implementation detail changes (the context shape/global store shape).

Radix usually treats context as *internal* so the expectation is that it is safe to change at will w/o introducing breaking changes. therefore, i'd be more concerned about that approach resulting in breaking changes for consumers. props are always the public facing API and a change there is expected to be a breaking change for all components. 

yes your understanding is correct. all packages wld always be on the same `DismissableLayer` version so only one version in bundle, and a change to props shape is the only scenario this issue can occur regardless of mix 'n' matched comp versions. 

a breaking change wld be handled the same way a breaking change is handled when props API changes for any of the comps (all packages bumped a major etc.). there is no need to create a new package for breaking changes. radix can just log in dev mode if `DismissableLayer` sniffs an old version of its public facing API being called.

```diff
type DismissableLayerProps {
- foo: boolean;
+ bar: boolean;
}
```
```tsx
const DismissableLayer = props => {
   if (props.foo !== undefined && process.NODE_ENV = 'development') {
    // stack trace wld likely show which comp used this prop
    throw new Error('A component is using an outdated version of DissmissableLayer. Please upgrade comp to latest');
   }
}
```

**Comment by jjenzz** [comment]
> Don't know if this will work for you, but I solved this in my library by using a Symbol: https://github.com/liqvidjs/plugins/blob/4f19efbd15399cecc1721142936b4f191de52d70/packages/playback/src/react.ts#L6-L14

i've just seen this btw and that is a neat solution ☺️ that wld work! it wld just suffer the issues i mentioned of breaking changes occurring when implementation detail changes, and consumer having multiple versions in their bundle unnecessarily.

**Comment by StephenHaney** [comment]
> > Don't know if this will work for you, but I solved this in my library by using a Symbol: https://github.com/liqvidjs/plugins/blob/4f19efbd15399cecc1721142936b4f191de52d70/packages/playback/src/react.ts#L6-L14
> 
> i've just seen this btw and that is a neat solution ☺️ that wld work! it wld just suffer the issues i mentioned of breaking changes occurring when implementation detail changes, and consumer having multiple versions in their bundle unnecessarily.

This seems like a good approach – do I understand correctly that using a symbol protects against collision with other libraries? Is there any known environment where this wouldn't work well?

**Comment by jjenzz** [comment]
perhaps i'm misunderstanding but as mentioned before, context shape is an implementation detail in radix. as far as i understand it, the symbol approach makes changes to an implementation detail become a breaking change for consumers. 

for example, if you have a context shape with a "foo" property in V1 and that property is removed in V2, an old comp could be looking for "foo" on the symbol's context but it won't exist.

tmk, the "latest" approach i proposed before makes changes to props (not context) the only potential breaking change. that is consistent with how changes in other primitives become breaking. props changes are the only potentially breaking changes for radix primitives tmk, since props are the public facing API. keeping that consistency seems more maintainable.

**Comment by leognmotta** [comment]
none of the solutions worked for me, even downgraded for testing select is unusable inside of a modal

**Comment by larissabakken** [comment]
> would still have to solve it the way radix suggests today but this would rarely be necessary because the api hardly ever changes.
@leognmotta 
Update the dependencies to –latest.

Try adding the overrides:

"overrides": {
  "@radix-ui/react-dismissable-layer": "^1.0.5",
  "@radix-ui/react-focus-scope": "^1.0.4"
}

Delete the yarn.lock, pnpm-lock.yaml, or package-lock.json file and the node_modules folder, then run the installation again.

https://github.com/shadcn-ui/ui/issues/1647

**Comment by kfalicov** [comment]
My problem was caused by these versions:

```
@radix-ui/react-dropdown-menu@2.1.3
@radix-ui/react-popover@1.1.3
@radix-ui/react-dialog@1.1.4
```

The latest version of `react-dropdown-menu` was `2.1.4`, and the latest version of `react-popover` was `1.1.4`. Updating to keep all 3 at `*.1.4` fixed my issue.

Is there a way for us to get installation warnings or a runtime warning for the mismatch? Is it safe to assume that matching minor versions means compatibility? Could any of this be fixed by using a non-locked peer dependency of `react-dismissable-layer` if not already?

**Comment by adrilasar** [comment]
Update @radix-ui worked for me also 👏

Update all command: `pnpm up "@radix-ui/*"`

**Comment by IPJT** [comment]
> [@jjenzz](https://github.com/jjenzz) I completely agree, I think we need to take a pragmatic approach here and remove the trap.

+1 
Just spent a a day debugging errors caused by this. Realised I did the same thing a year back... Having some errors in dev seems like the obvious quick-fix.

**Comment by KinanGH98** [comment]
#1748

> My problem was caused by these versions:
> 
> ```
> @radix-ui/react-dropdown-menu@2.1.3
> @radix-ui/react-popover@1.1.3
> @radix-ui/react-dialog@1.1.4
> ```
> 
> The latest version of `react-dropdown-menu` was `2.1.4`, and the latest version of `react-popover` was `1.1.4`. Updating to keep all 3 at `*.1.4` fixed my issue.
> 
> Is there a way for us to get installation warnings or a runtime warning for the mismatch? Is it safe to assume that matching minor versions means compatibility? Could any of this be fixed by using a non-locked peer dependency of `react-dismissable-layer` if not already?

Thanks for this fix! The versions that are currently working together are:
```
"@radix-ui/react-dropdown-menu": "2.1.4",
"@radix-ui/react-select": "2.1.4",
"@radix-ui/react-popover": "1.1.4",
"@radix-ui/react-dialog": "1.1.4",

"overrides": {
  "@radix-ui/react-dismissable-layer": "1.1.4"
}
```

**Comment by avinean** [comment]
Hello people!
I had the same issue.
1. open dialog
2. open select
3. click outside the dialog
dialog is closed but pointer-events: none remains.

The issue has gone after I removed node_modules and reinstalled packages.

"@radix-ui/react-dialog": "^1.1.14",
"@radix-ui/react-select": "^2.2.5",
and as their dependency
"@radix-ui/react-dismissable-layer" "1.1.11"

**Comment by chaance** [comment]
This shouldn't be an issue moving forward as our recommendation now is to use the consolidated `radix-ui` package.

**Comment by kfalicov** [comment]
> This shouldn't be an issue moving forward as our recommendation now is to use the consolidated `radix-ui` package.

As one of the primary strengths of Radix as a component library over alternatives such as Mantine or Mui has always been the modularity of it, what can we expect about the LTS lifetime of the independent modules?  Is there a blog post or announcement about this recommendation?

If the plans are always to maintain the individual primitives alongside the consolidated package, then I have no issue
